### ISSUE TITLE
Spaces in S3 object keys

### ISSUE THREAD

**Issue by jayroh** [taylorfinnell/awscr-s3#34]
Hey there! Love the shard, so thank you _very_ much. It's fantastic.

Question for you - have you run into issues uploading where the target S3 object key has spaces in it? When I attempt to do that I get the following error: 

```
HttpVersionNotSupported: The HTTP version specified is not supported. (Awscr::S3::Http::ServerError)
```

Which is a little odd in itself. When I change the key's spaces to underscores, however - all is well. Thanks so much!

**Comment by taylorfinnell** [comment]
You're welcome! I can't look into in detail atm but I think it may be related to the request signing.

https://github.com/taylorfinnell/awscr-signer

I can't remember exactly why the spaces we're an issue, it may have been a crystal limitation. I can look into it more in a couple days :)

**Comment by jayroh** [comment]
Right on man, thank you! I'll dig around a little too as I explore this
weekend

😁

**Comment by taylorfinnell** [comment]
@jayroh 

I think this should take care of it at least if you're using v4 signatures (the default). https://github.com/taylorfinnell/awscr-signer/pull/34

Basically, the signature was being calculated correctly (as the path was properly escaped in signature calculation), but amazon was still seeing the request as:

```
GET /hello world HTTP/1.1
```

And not know that the world was part of the path. So it needed to be escaped. We can't just escape before calculating signature because then the `%20` will get escaped again during signature calcuation. Going to look into fixing it for both signing versions, and see if I can come up with something better before I merge this

**Comment by jayroh** [comment]
Don't you just love the signature creation process? I just recently went through the v2 -> v4 signing process for an app I manage and at times felt like I missed a "sacrifice chicken" step somewhere.

In any case, thanks for taking a look at this, dude. I appreciate it. If there's any part that could use an extra hand feel free to point me in the general direction and I'll help out.

**Comment by jrester** [comment]
@taylorfinnell 
Maybe it would be better to encode the object name before signing and skip encoding the full path in the signing process. This would also be helpful in cases where the name includes `/`. I guess the method `URI#encode_www_form` would do the job. 
This would also fix #60
